### ISSUE TITLE
[BUG] Generally buggy behaviour when proxying Nextcloud

### ISSUE THREAD

**Issue by CorneliusCornbread** [tobychui/zoraxy#164]
So I'm attempting to proxy my nextcloud instance, I've got the trusted IPs setup in nextcloud and have had this nextcloud instance proxied successfully via NPM in the past. However, it seems like Nextcloud really doesn't like being proxied via Zoraxy. Looking at the browser console, it seems like it's not properly proxying resource requests to be done via the domain, rather trying to directly access the resources via either the docker IP or the local IP address.

![image](https://github.com/tobychui/zoraxy/assets/8294697/a6787e3a-250d-4722-ab37-1cdf6207a336)

**To Reproduce**
Nextcloud proxy:
![image](https://github.com/tobychui/zoraxy/assets/8294697/2be64e8e-7641-40c1-a4e3-d8ece33f5765)
Changing to the local IP address instead of the docker one will simply result in it working, but only VIA LAN connections.

**Expected behavior**
Nextcloud should load as normal.

**Browser (if it is a bug appears on the UI section of the system):**
 - OS: Fedora (tested non LAN connections via phone data on Android)
 - Browser: Firefox
 - Version: 126.0

**Host Environment (please complete the following information):**
-  Arch: x86
 - Device: Ryzen 3600x system
 - OS: Unraid
 - Version: 6.12.10

**Additional context**
I cannot reproduce this behaviour with my Crafty or Dozzle instance, they work just fine.

**Comment by tobychui** [comment]
I guess either your network setting have issues or your NextCloud setting is wrong. I don't use NextCloud, but this similar issue sometime happens to Wordpress too, which was caused by invalid settings of based URL / domains.
As you can see from the first part of the error message, your next cloud web interface trying to load the resources (svg) from its docker IP address
![圖片](https://github.com/tobychui/zoraxy/assets/24617523/fa5d123b-599a-47c2-8594-2d676389ca3f)

But from the later part, I assume you are connecting to it via local IP address, due to `self` src requirement, it cannot load as both address (docker IP start with 172 and your connecting IP start with 192) doesn't match.
![圖片](https://github.com/tobychui/zoraxy/assets/24617523/e956ad98-6098-46b8-b5fc-bd17a06e6295)

This is a common issue with apps that didn't design to be reverse proxied. So to make it correctly load, you should be able to set the base URL / domain like wordpress that point to your (sub)domain name and force NextCloud to load resources from your (sub)domain URL instead of 172.

**Comment by Morethanevil** [comment]
If you are using Nextcloud in docker, use the normal HTTP Port (80). SSL is done via Zoraxy, uncheck "Require TLS" and "Skip verififcation" and uncheck "Allow Plain HTTP"

Require TLS and Allow plain HTTP are exactly the opposite :)

 I have Nextcloud running on my host. Zoraxy points to my Apache webserver which is serving Nextcloud. It works on the normal HTTP Port. Certificates are handled by Zoraxy, you should avoid additional HTTPS ports or configuration.


Do you use the AIO image from Nextcloud?

**Comment by nettybun** [comment]
Nextcloud is getting confused because Zoraxy is sending the HTTP Host header of the endpoint, not the domain. If you configure Zoraxy to route `"nextcloud.example.com"` to `"127.0.0.1:8443"` you'll get HTTP `Host` header as 127.0.0.1 and HTTP `Referrer` header as https://127.0.0.1:8443

I'm not sure if this is by design, but Nextcloud doesn't like it.

The image I have is `lscr.io/linuxserver/nextcloud:latest` which uses nginx+php-fpm. I'm also running Podman rootless as the container engine.

Here are the nginx error.log and access.log:
```
root@c8611a770e26:/config/log/nginx# tail error.log 
2024/05/23 00:43:46 [error] 392#392: *3 upstream timed out (110: Operation timed out) while reading response header from upstream, client: 10.88.0.11, server: _, request: "POST / HTTP/2.0", upstream: "fastcgi://127.0.0.1:9000", host: "127.0.0.1:8443"
root@c8611a770e26:/config/log/nginx# tail access.log 
10.88.0.13 - - [23/May/2024:01:00:55 -0700] "GET / HTTP/2.0" 302 0 "-" "Go-http-client/2.0"
10.88.0.13 - - [23/May/2024:01:00:55 -0700] "GET /index.php/login HTTP/2.0" 200 14824 "https://127.0.0.1:8443" "Go-http-client/2.0"
10.88.0.13 - - [23/May/2024:01:05:55 -0700] "GET / HTTP/2.0" 302 0 "-" "Go-http-client/2.0"
10.88.0.13 - - [23/May/2024:01:05:55 -0700] "GET /index.php/login HTTP/2.0" 200 14832 "https://127.0.0.1:8443" "Go-http-client/2.0"
```

The `Go-http-client` is Zoraxy.

The container image I'm using has this Docker-specific line:
```
    # display real ip in nginx logs when connected through reverse proxy via docker network
    set_real_ip_from 172.16.0.0/12;
    real_ip_header X-Forwarded-For;

```
I changed that CIDR from Docker's 172.x to Podman's `10.88.0.0/16` and then suddenly my logs were fixed:
```
root@c8611a770e26:/config/log/nginx# tail access.log 
107.11.32.102 - - [23/May/2024:01:55:12 -0700] "GET / HTTP/2.0" 302 0 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:124.0) Gecko/20100101 Firefox/124.0"
107.11.32.102 - - [23/May/2024:01:55:13 -0700] "GET /index.php/login HTTP/2.0" 200 14832 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:124.0) Gecko/20100101 Firefox/124.0"
``` 

Still, the buggy broken links remain in the UI.

I think it's this fastcgi param line that is giving PHP the wrong server name?
```
root@c8611a770e26:/# nginx -T | grep fastcgi_param
...
fastcgi_param  SERVER_NAME        $host; # Send HTTP_HOST as SERVER_NAME. If HTTP_HOST is blank, send the value of server_name from nginx (default is `_`)
```

I asked nginx to log its `$http_host` and `$host` to confirm Zoraxy is sending the wrong header:
```
    # Sets the path, format, and configuration for a buffered log write.
    log_format hostcombined '$remote_addr - $remote_user [$time_local] '
                    '"$request" $status $body_bytes_sent '
                    'HOST:"$host" HTTP_HOST:"$http_host" REF:"$http_referer" "$http_user_agent"';
    access_log /config/log/nginx/access.log hostcombined;
```

It seems:
```
107.11.32.102 - - [23/May/2024:02:17:08 -0700] "PUT /ocs/v2.php/apps/user_status/api/v1/heartbeat?format=json HTTP/2.0" 200 149 HOST:"127.0.0.1" HTTP_HOST:"127.0.0.1:8443" REF:"-" "Mozilla/5.0 (X11; Linux x86_64; rv:124.0) Gecko/20100101 Firefox/124.0"
107.11.32.102 - - [23/May/2024:02:17:14 -0700] "GET /ocs/v2.php/apps/notifications/api/v2/notifications HTTP/2.0" 304 0 HOST:"127.0.0.1" HTTP_HOST:"127.0.0.1:8443" REF:"-" "Mozilla/5.0 (X11; Linux x86_64; rv:124.0) Gecko/20100101 Firefox/124.0"
```

So to workaround this in Nextcloud (or any downstream app that relies on HTTP Host header) you'll have to hardcode the nginx/apache server config to your domain... (note: nginx does not let you modify `$host`!)

```
fastcgi_param SERVER_NAME nextcloud.example.com
```
Fixed it for me! :+1:

**Comment by tobychui** [comment]
@nettybun Thanks for the details debug and fixing process!
I will try to test this out later this weekend and see if I need to introduce a new switch for automatic hostname rewriting based on inbound domain name.

**Comment by tobychui** [comment]
That is interesting, as I am testing Zoraxy natively on my system (Windows) with a debug php script, I get the followings.
```php
<?php
$hostname = $_SERVER['HTTP_HOST'];
echo $hostname;
echo '<br>';

$hostname = $_SERVER['SERVER_NAME'];
echo $hostname;
echo '<br>';

$serverIp = $_SERVER['SERVER_ADDR'];
echo $serverIp;
echo '<br>';
```
Results
```
test.imuslab.internal
test.imuslab.internal
127.0.0.1
```

So php indeed grab the correct HTTP_HOST sent from Zoraxy. Besides, the `Go-http-client` should not be coming from Zoraxy, as in Zoraxy source code, the user agent information was removed from the outgoing request to the back-end server in `dpcore.go`
```go
//Hide Go-HTTP-Client UA if the client didnt sent us one
	if _, ok := header["User-Agent"]; !ok {
		// If the outbound request doesn't have a User-Agent header set,
		// don't send the default Go HTTP client User-Agent.
		header.Set("User-Agent", "")
	}
```

My assumption is that there is something wrong between Zoraxy and the Nginx server running NextCloud (php) and I cannot reproduce it with my environment. If anyone is available, please help by use the attached debug.php file and provide more information regarding setups that have shows this bug.
[debug.zip](https://github.com/tobychui/zoraxy/files/15415860/debug.zip)

**Comment by nettybun** [comment]
@tobychui whoa! Thanks for putting so much effort into it. However, I worry you may be seeing the same issue as me:

![Screenshot_20240523-095134](https://github.com/tobychui/zoraxy/assets/44614862/3502c71e-46a4-4e71-9a73-d3d5b02bde81)

If you hover on these thumbnails it will say http://192.168...? Maybe?

I'll be back at a laptop soon and I can test with your debug script.

**Comment by nettybun** [comment]
![image](https://github.com/tobychui/zoraxy/assets/44614862/552ae2ee-58d1-4308-8fbb-753dafa76f8d)

I used your debug.php script by directly putting it into the nextcloud image under `/app/www/public` (but nextcloud disallows unknown php files since that's dangerous? I think? so I had to re-use/modify an existing file; I chose status.php)

Seems like the HTTP Host header is mangled somehow very odd... I'll keep investigating!

**Comment by nettybun** [comment]
![image](https://github.com/tobychui/zoraxy/assets/44614862/09ebff40-d48b-4ca6-8919-e54ea01ab712)

This looks like the cause of the bug. @tobychui I don't understand the code comment; can you advise if it is safe to modify this?

I modify the code:
![image](https://github.com/tobychui/zoraxy/assets/44614862/d9ec240c-1099-4902-84f4-f01bdbbf1486)

Rebuild with `go build`.

Yes this is the issue:
![image](https://github.com/tobychui/zoraxy/assets/44614862/3f7bf5d9-93e8-4bfa-a72a-e415cfd7b35e)

I modify the code again:
![image](https://github.com/tobychui/zoraxy/assets/44614862/a4894b25-55c3-411d-b303-b83e8d26e177)

Problem fixed!
![image](https://github.com/tobychui/zoraxy/assets/44614862/4f471031-a47f-498d-b861-99af09325824)

This also fixed Nextcloud! See how now the thumbnails are working - they aren't "?" anymore, they are the real thing:
![image](https://github.com/tobychui/zoraxy/assets/44614862/7f75bc62-4f99-4af5-b305-1f951eb5f6a6)

Can we push this fix?


I worry the code comment is telling me the change is not this easy though >_>

**Comment by tobychui** [comment]
After adding the domain in the config.php inside the docker volume and patched PR #168, this bug has been fixed.
![圖片](https://github.com/tobychui/zoraxy/assets/24617523/59608d1c-f62f-43b5-8e66-43fced9500e7)

I guess this bug also partly fixed the Promox issues which I will investigate later. Thanks so much for everyone involved and help making Zoraxy a better tool!

**Comment by amura11** [comment]
I know this is an older post but I came across this while trying to figure out why my Nextcloud was behaving weirdly for me and thought I'd share my fix in case it can help someone in the future. A lot of the other fixes I found only worked partially for me.

## The Fix

Open `/<Netxcloud Docker Folder>/nginx/site-confs/default.conf` and comment out these lines (lines 23 and 24 for me):
```
    set_real_ip_from 172.16.0.0/12;
    real_ip_header X-Forwarded-For;
```

Then add your Zoraxy container's IP address (or the whole docker network) to the `trusted_proxies` entry in the `config.php`, Eg:
```
  'trusted_proxies' =>
  array (
    0 => '172.20.0.0/16',
  ),
```

Restart and Nextcloud should behave. I'm not sure if this will have other affects but so far all I've noticed is the logs only show the Zoraxy container IP as opposed to the client IPs.

What I was seeing was that Nextcloud would ignore the proxy headers if the IP it saw wasn't in the trusted proxies. The problem is that the IP it was seeing was the client's real IP which could be something completely outside my network for external access. I found that putting those IPs in the `trusted_proxies` list did work but these IPs will change. It seems that removing those two lines makes Nextcloud always see the Zoraxy IP which can be added to the `trusted_proxies`.
